src/nectarchain/makers/component/charges_component.py

```diff
@@ -568,44 +566,42 @@ def compute_charges(
         imageExtractor = __class__._get_imageExtractor(
             method=method, subarray=subarray, **kwargs
         )
-        if channel == constants.HIGH_GAIN:
-            out = np.array(
-                [
-                    CtapipeExtractor.get_image_peak_time(
-                        imageExtractor(
-                            waveformsContainer.wfs_hg[i],
-                            tel_id,
-                            channel,
-                            waveformsContainer.broken_pixels_hg[i],
-                        )
+        out = np.array(
+            [
+                CtapipeExtractor.get_image_peak_time(
+                    imageExtractor(
+                        waveforms=np.array(
+                            [waveformsContainer.wfs_hg[i], waveformsContainer.wfs_lg[i]]
+                        ),
+                        tel_id=tel_id,
+                        selected_gain_channel=None,
+                        broken_pixels=np.array(
+                            [
+                                waveformsContainer.broken_pixels_hg[i],
+                                waveformsContainer.broken_pixels_lg[i],
+                            ]
+                        ),
                     )
-                    for i in range(len(waveformsContainer.wfs_hg))
-                ]
-            ).transpose(1, 0, 2)
-            return ChargesContainer.fields["charges_hg"].dtype.type(
-                out[0]
-            ), ChargesContainer.fields["peak_hg"].dtype.type(out[1])
+                )
+                for i in range(len(waveformsContainer.wfs_hg))
+            ]
+        )
+        # by setting selected_gain_channel to None, we can now pass to ctapipe extractor
+        # waveforms in (n_ch, n_pix, n_samples)
+        # then out hase shape (n_events, 2, n_ch, n_pi)
+        if channel == constants.HIGH_GAIN:
+            index = 0
+            gain_label = "hg"
         elif channel == constants.LOW_GAIN:
-            out = np.array(
-                [
-                    CtapipeExtractor.get_image_peak_time(
-                        imageExtractor(
-                            waveformsContainer.wfs_lg[i],
-                            tel_id,
-                            channel,
-                            waveformsContainer.broken_pixels_lg[i],
-                        )
-                    )
-                    for i in range(len(waveformsContainer.wfs_lg))
-                ]
-            ).transpose(1, 0, 2)
-            return ChargesContainer.fields["charges_lg"].dtype.type(
-                out[0]
-            ), ChargesContainer.fields["peak_lg"].dtype.type(out[1])
+            index = 1
+            gain_label = "lg"
         else:
             raise ArgumentError(
                 None, f"channel must be {constants.LOW_GAIN} or {constants.HIGH_GAIN}"
             )
+        return ChargesContainer.fields[f"charges_{gain_label}"].dtype.type(
```

Comment:
Here we are computing both for low and high gain channel de charge, however the function return only one of these quantities. It's more elegant, but I'm wondering what is lost in term of performance. (Actually the charge extraction is `numba` based so I guess it is negligible).

Reply:
Since this method has a mandatory argument for the channel, I'd say it may be more efficient to only compute the charge in the requested channel.

Reply:
Hi @jlenain, yes, I think it's perfect like this, and it avoids the loop over waveforms